src/framework/net/webconnection.cpp

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2024 OTArchive <https://otarchive.com>
+ * Copyright (c) 2010-2022 OTClient <https://github.com/edubart/otclient>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifdef __EMSCRIPTEN__
+
+#include "webconnection.h"
+
+#include <framework/core/application.h>
+#include <client/game.h>
+
+#include <utility>
+#include <asio/read.hpp>
+#include <asio/read_until.hpp>
+#include <framework/core/eventdispatcher.h>
+
+#include <emscripten/threading.h>
+
+
+
+std::list<std::shared_ptr<asio::streambuf>> WebConnection::m_outputStreams;
+WebConnection::WebConnection()
+{
+}
+
+WebConnection::~WebConnection()
+{
+#ifndef NDEBUG
+    assert(!g_app.isTerminated());
+#endif
+    close();
+}
+
+void WebConnection::poll()
+{
+
+}
+
+void WebConnection::terminate()
+{
+    emscripten_websocket_deinitialize();
+}
+
+void WebConnection::close()
+{
+    if (!m_connected && !m_connecting)
+        return;
+
+    emscripten_websocket_deinitialize();
+    m_websocket = 0;
+
+    if (!m_gameWorld)
+        return;
+
+    m_connecting = false;
+    m_connected = false;
+    m_connectCallback = nullptr;
+    m_errorCallback = nullptr;
+    m_recvCallback = nullptr;
+
+    m_inputStream.consume(m_inputStream.size());
+
+    //Workaround for the abrupt termination of the websocket
+    g_dispatcher.addEvent([] {
+        g_game.forceLogout();
+    });
+}
+
+
+void WebConnection::connect(const std::string_view host, uint16_t port, const std::function<void()>& connectCallback, bool gameWorld)
+{
+    m_gameWorld = gameWorld;
+
+    m_connected = false;
+    m_connecting = true;
+    m_connectCallback = connectCallback;
+
+    std::string ip = std::string(host);
+    ip.length() == 0 ? ip = "localhost" : ip;
+
+#ifndef NDEBUG
+    const std::string prefix = "ws://";
+#else
+    const std::string prefix = "wss://";
+#endif
```

Comment:
```suggestion
#ifdef NDEBUG
    const std::string prefix = "wss://";
#else
    const std::string prefix = "ws://";
#endif
```
the double negation thing can be confusing 😅  this is easier to read IMO